### PR TITLE
fix(typescript-estree): hash code to reduce update frequency

### LINKED ISSUE #1132: [2.5.0] Lint runs are ~30% slower
Spinning off of discussion in #1126.
Reported by: @doberkofler

I think that this is because of the change in #1083 

There are use cases like Vue, or Markdown, wherein we get provided N distinct content strings for the same filename, where N is the number of `<script>` tags in the vue file, or the number of code blocks in a markdown file.

We used to not trigger a program sync if we hadn't been asked to parse the file yet, but for these use cases, the file contents TS reads are always different to the file contents eslint gives us, so we have to trigger a file sync on the first parse for a file.

I'll need to look at a solution for this, as invalidating the program for every single file in a CLI run for pure JS/TS codebases (when the file contents from eslint === the file contents in ts) is pretty silly.

### PR DESCRIPTION
**Pull request by yoyo930021**
try to fix #1132

### CODE REVIEW

**Comment by yoyo930021** [comment]
We need to look at another solution for Vue SFC is slow.

**Comment by yoyo930021** [comment]
Test case (use https://github.com/typescript-eslint/vega-lite):
- v2.6.1:
```
real 10.40
user 15.49
sys 0.80
```
- This PR:
```
real 6.62
user 10.21
sys 0.61
```
- v2.3.1:
```
real 6.32
user 10.03
sys 0.58
```

File: packages/typescript-estree/src/create-program/createWatchProgram.ts [hunk 1]
```diff
@@ -104,6 +104,13 @@ function diagnosticReporter(diagnostic: ts.Diagnostic): void {
   );
 }
 
+function createHash(content: string): string {
+  if (ts.sys.createHash) {
```
Comment (bradzacher) [inline packages/typescript-estree/src/create-program/createWatchProgram.ts]: `ts.sys` won't exist in browser environments.

```suggestion
  if (ts.sys && ts.sys.createHash) {
```

File: packages/typescript-estree/src/create-program/createWatchProgram.ts [hunk 2]
```diff
@@ -125,10 +132,11 @@ function getProgramsForProjects(
   currentLintOperationState.filePath = filePath;
 
   // Update file version if necessary
-  // TODO: only update when necessary, currently marks as changed on every lint
   const fileWatchCallbacks = fileWatchCallbackTrackingMap.get(filePath);
+  const codeHash = createHash(code);
   if (
-    parsedFilesSeen.has(filePath) &&
+    parsedFilesSeenHash.has(filePath) &&
+    parsedFilesSeenHash.get(filePath) !== codeHash &&
```
Comment (bradzacher) [inline packages/typescript-estree/src/create-program/createWatchProgram.ts]: No point doing a `has` followed by a `get`

```ts
const existingCodeHash = parsedFilesSeenHash.get(filePath);
if (
  existingCodeHash &&
  existingCodeHash !== codeHash
```

**Comment by DanceParty** [comment]
Can't wait for this to be merged. This PR seems to bring our webpack build times from ~10 minutes to ~2 minutes and my node cpu usage at idle from ~150% to ~0%. Thanks 🙏

**Comment by DanceParty** [comment]
Cannot wait to have this merged, my built time is drastically quicker and node cpu is much lower using this change. Thanks :tada:

File: packages/typescript-estree/src/create-program/createWatchProgram.ts (hunk 1 above)
Comment (yoyo930021) [reply packages/typescript-estree/src/create-program/createWatchProgram.ts]: I don't know about this.

**Comment by JoshuaKGoldberg** [comment]
FWIW, _(presumably this fix in)_ 2.7.0 brought time to lint a ~2000 file, 90% TS / 10% JS project from 10 minutes to 2. Awesome fix, thank you so much! 🚀

**Pull request merged**